trade-finance-manager-ui/server/controllers/login/loginService.js

```
@@ -0,0 +1,72 @@
+const { SSO } = require("../../constants");
+
+/**
+ * verifyReferrerForExternalSsoPost
+ * Verify referrer, allow empty referrer for localhost.
+ * 1) If host is localhost then referrer will be missing, no need to check it
+ * 2) Return error message if referrer is missing or it is not from SSO authority
+ * @param {Object} req: Express request
+ * @returns {Boolean|String} False or error message.
+ */
+const verifyReferrerForExternalSsoPost = (req) => {
+  const hostnameWithoutPort = req.get('host').split(':')[0];
+  const referrer = req.get('Referrer');
+
+  // Referrer is not available in localhost because of policy "no-referrer-when-downgrade".
+  if (hostnameWithoutPort === 'localhost') {
+    return false;
+  }
+
+  if (referrer && referrer.indexOf(`${SSO.AUTHORITY}/`) === 0) {
+    return false;
+  }
+
+  console.error('Login request comming from unexpected website: %s', referrer);
+  return 'Login request comming from unexpected website.';
+}
+
+/**
+ * verifyBodyForExternalSsoPost
+ * Verify SSO input parameters, allow limited/safe number of characters.
+ * @param {Object} req: Express request
+ * @returns {Boolean|String} False or error message.
+ */
+const verifyBodyForExternalSsoPost = (templateParams) => {
```

Again I think this should throw an error, what do you think?